### PR TITLE
[Feature] New RBAC privilege framework: role inheritance + set role

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Introduce a new privilege framework.
To avoid huge PR which is hard to review and easy to conflict, I'm splitting into several small PR, each implementing an individual aspect of the new privilege framework. This is the tenth one. For previous PRs see 
https://github.com/StarRocks/starrocks/pull/11447
https://github.com/StarRocks/starrocks/pull/11528
https://github.com/StarRocks/starrocks/pull/11593
https://github.com/StarRocks/starrocks/pull/11732
https://github.com/StarRocks/starrocks/pull/11777
https://github.com/StarRocks/starrocks/pull/11847
https://github.com/StarRocks/starrocks/pull/11909
https://github.com/StarRocks/starrocks/pull/12035
https://github.com/StarRocks/starrocks/pull/12128

This PR implements some advanced usage on roles, including
* Allow granting/revoking one role to another, with necessary checks such as inheritance depth & circle detection.
* Allow using `SET ROLE` statement to explicitly use roles to the current user.
* Add restriction on all possible roles on one specific user

This PR also makes some relevant changes due to the above features.
* User privilege collection will be deleted immediately after the user is dropped.
* Asynchronous cleanup on roles & objects is refactored to avoid deadlock.
* `RolePrivilegeCollectionInfo` contains multiple role privilege collections instead of one so that both the parent role and the subrole can be updated and recorded in one journal.
* Add explicitly postfix `Unlocked` to all interfaces in PrivilegeManager that are public or protected and visit critical resources without any locks.

What's left:
* Default roles & build-in roles are still under development.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 40 / 66 (60.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[660]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|5|00.00%|[491, 492, 916, 917, 918]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|0|3|00.00%|[276, 280, 281]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|0|1|00.00%|[194]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeRoleStmt.java|1|3|33.33%|[16, 17]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|8|14|57.14%|[3680, 3681, 3682, 3694, 3695, 3696]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantRoleStmt.java|2|3|66.67%|[16]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeRoleStmt.java|10|15|66.67%|[25, 26, 27, 28, 51]|
|:large_blue_circle:|com/starrocks/sql/ast/SetRoleStmt.java|7|8|87.50%|[31]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|9|10|90.00%|[106]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 43 / 69 (62.32%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[660]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|5|00.00%|[491, 492, 916, 917, 918]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|0|3|00.00%|[276, 280, 281]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeRoleStmt.java|1|3|33.33%|[16, 17]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|1|2|50.00%|[193]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|8|14|57.14%|[3680, 3681, 3682, 3694, 3695, 3696]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantRoleStmt.java|2|3|66.67%|[16]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeRoleStmt.java|10|15|66.67%|[25, 26, 27, 28, 51]|
|:large_blue_circle:|com/starrocks/sql/ast/SetRoleStmt.java|7|8|87.50%|[31]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|9|10|90.00%|[106]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 260 / 278 (93.53%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[660]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|5|00.00%|[491, 492, 916, 917, 918]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|0|3|00.00%|[276, 280, 281]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|4|50.00%|[928, 933]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|5|6|83.33%|[193]|
|:large_blue_circle:|com/starrocks/sql/ast/SetRoleStmt.java|7|8|87.50%|[31]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|9|10|90.00%|[106]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeRoleStmt.java|14|15|93.33%|[51]|
|:large_blue_circle:|com/starrocks/privilege/RolePrivilegeCollection.java|17|18|94.44%|[89]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|173|175|98.86%|[397, 670]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/UserPEntryObject.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationManager.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 308 / 319 (96.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|4|50.00%|[928, 933]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|5|9|55.56%|[192, 193, 194, 195]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|9|10|90.00%|[106]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeRoleStmt.java|14|15|93.33%|[51]|
|:large_blue_circle:|com/starrocks/qe/SetRoleExecutor.java|21|22|95.45%|[14]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|190|192|98.96%|[397, 671]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/UserPEntryObject.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/RolePrivilegeCollection.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationManager.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/SetRoleStmt.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by padmejin** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 313 / 322 (97.20%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|2|4|50.00%|[916, 917]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|4|50.00%|[928, 933]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|9|10|90.00%|[106]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeRoleStmt.java|14|15|93.33%|[51]|
|:large_blue_circle:|com/starrocks/qe/SetRoleExecutor.java|21|22|95.45%|[14]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|191|193|98.96%|[397, 675]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/UserPEntryObject.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/RolePrivilegeCollection.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationManager.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/SetRoleStmt.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|10|10|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java [hunk 1]
```diff
@@ -670,20 +856,137 @@ public PEntryObject analyzeObject(String type, List<String> allTypes, String res
...
+                }
+            } finally {
+                roleReadUnlock();
+            }
+        } finally {
+            userWriteUnlock();
+        }
+
+        // 3. remove invalidate object of roles
+        // we have to add user lock first because it may contains user privilege
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java:890]: why we need user lock, i don't understand the comment very much.

**Review by nshangyiming — DISMISSED** [review]
approved

File: fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java [hunk 2]
```diff
@@ -908,6 +912,11 @@ private void handleExecAsStmt() {
         ExecuteAsExecutor.execute((ExecuteAsStmt) parsedStmt, context);
     }
 
+    private void handleSetRole() throws PrivilegeException, UserException {
+        SetRoleExecutor.execute((SetRoleStmt) parsedStmt, context);
+
```
[the remark this answers is not in the document]
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java [hunk 3]
```diff
@@ -75,18 +75,17 @@ private void analyseUser(UserIdentity userIdent, ConnectContext session, boolean
...
+        private void analyseRoleName(String roleName, ConnectContext session, boolean canBeAdmin, String errMsg) {
+            validRoleName(roleName, canBeAdmin, errMsg);
+            if (!session.getGlobalStateMgr().getAuth().doesRoleExist(roleName)) {
+                throw new SemanticException("role " + roleName + " not exist!");
             }
-            return qualifiedRole;
         }
 
         /**
          * check if role name valid and get full role name
```
[the remark this answers is not in the document]
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java]: done

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java [hunk 4]
```diff
@@ -286,44 +288,157 @@ protected void revokeFromRole(
...
+                throw new PrivilegeException(String.format("role %s[%d] is already a predecessor role of %s[%d]",
+                        roleName, roleId, parentRoleName, parentRoleId));
+            }
+
+            // temporarily add sub role to parent role to verify inheritance depth
+            boolean verifyDone = false;
+            parentCollection.addSubRole(roleId);
+            try {
+                // verify role inheritance depth
+                parentRolePrecessors = getAllPredecessorsUnlocked(parentRoleId);
```
[the remark this answers is not in the document]
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java]: done

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 313 / 322 (97.20%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|2|4|50.00%|[916, 917]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|4|50.00%|[928, 933]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|9|10|90.00%|[106]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeRoleStmt.java|14|15|93.33%|[51]|
|:large_blue_circle:|com/starrocks/qe/SetRoleExecutor.java|21|22|95.45%|[14]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|191|193|98.96%|[397, 675]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/UserPEntryObject.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/RolePrivilegeCollection.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationManager.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/SetRoleStmt.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|10|10|100.00%|[]|

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 314 / 321 (97.82%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|4|50.00%|[928, 933]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|9|10|90.00%|[106]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeRoleStmt.java|14|15|93.33%|[51]|
|:large_blue_circle:|com/starrocks/qe/SetRoleExecutor.java|21|22|95.45%|[14]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|191|193|98.96%|[397, 675]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/UserPEntryObject.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/RolePrivilegeCollection.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationManager.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/SetRoleStmt.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|10|10|100.00%|[]|

**Comment by padmejin** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java (hunk 1 above)
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java:890]: got it

**Review by nshangyiming — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by gengjun-git — APPROVED** [review]

**Pull request merged**